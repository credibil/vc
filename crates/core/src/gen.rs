//! # Generate
//!
//! Generate random strinsg for use in authorzation code, token, state,
//! nonce, etc.

// LATER: replace with a proper random string generator

/// Random string generation for auth code, token, state, and nonce.
use base64ct::{Base64UrlUnpadded, Encoding};

const PIN_CHARS: &str = "0123456789";
const PIN_LEN: usize = 6;

// "'`+=,./\|:;?><}{][_-
const SAFE_CHARS: &str =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789)(*&^%$#@!~";
const STATE_LEN: usize = 32;

/// Generates a base64 encoded random string for authorization code.
#[must_use]
pub fn auth_code() -> String {
    let rnd = random_string(STATE_LEN, SAFE_CHARS);
    Base64UrlUnpadded::encode_string(rnd.as_bytes())
}

/// Generates a base64 encoded random string for token
#[must_use]
pub fn token() -> String {
    let rnd = random_string(STATE_LEN, SAFE_CHARS);
    Base64UrlUnpadded::encode_string(rnd.as_bytes())
}

/// Generates a base64 encoded random string for deferred issuance
/// `transaction_id`.
#[must_use]
pub fn transaction_id() -> String {
    let rnd = random_string(STATE_LEN, SAFE_CHARS);
    Base64UrlUnpadded::encode_string(rnd.as_bytes())
}

/// Generates a base64 encoded random string for deferred issuance
/// `transaction_id`.
#[must_use]
pub fn notification_id() -> String {
    let rnd = random_string(STATE_LEN, SAFE_CHARS);
    Base64UrlUnpadded::encode_string(rnd.as_bytes())
}

/// Generates a base64 encoded random string for nonce
#[must_use]
pub fn nonce() -> String {
    let rnd = random_string(STATE_LEN, SAFE_CHARS);
    Base64UrlUnpadded::encode_string(rnd.as_bytes())
}

/// Generates a base64 encoded random string for `issuer_state`
#[must_use]
pub fn state_key() -> String {
    let rnd = random_string(STATE_LEN, SAFE_CHARS);
    Base64UrlUnpadded::encode_string(rnd.as_bytes())
}

/// Generates a use PIN code.
#[must_use]
pub fn tx_code() -> String {
    random_string(PIN_LEN, PIN_CHARS)
}

// Generates a random string from a given set of characters. Uses fastrand so is
// not cryptographically secure.
fn random_string(len: usize, charset: &str) -> String {
    let chars: Vec<char> = charset.chars().collect();
    (0..len).map(|_| chars[fastrand::usize(..chars.len())]).collect()
}

// ///
// /// Generate a new random, base64-encoded 128-bit CSRF token.
// pub fn new_random() -> Self {
//     CsrfToken::new_random_len(16)
// }
// ///
// /// Generate a new random, base64-encoded CSRF token of the specified length.
// ///
// /// # Arguments
// ///
// /// * `num_bytes` - Number of random bytes to generate, prior to
// base64-encoding. pub fn new_random_len(num_bytes: u32) -> Self {
//     let random_bytes: Vec<u8> = (0..num_bytes).map(|_|
// thread_rng().gen::<u8>()).collect();
//     CsrfToken::new(base64::encode_config(&random_bytes,
// base64::URL_SAFE_NO_PAD)) }
