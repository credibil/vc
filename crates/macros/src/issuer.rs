use proc_macro2::{Span, TokenStream};
use quote::quote;
use syn::parse::{Error, Result};

use crate::parse::Data;

pub fn create_offer_request(input: &Data) -> Result<TokenStream> {
    // remove fields so we can error on any remaining
    let mut fields = input.fields.clone();
    let credential_issuer = fields.remove("credential_issuer").unwrap();
    let subject_id = fields.remove("subject_id").unwrap();
    let credential_configuration_ids = fields.remove("credential_configuration_ids").unwrap();
    let pre_authorize = fields.remove("pre-authorize").unwrap();
    let tx_code_required = fields.remove("tx_code_required").unwrap();
    let send_type = fields.remove("send_type").unwrap();

    // return error for any additional fields
    if !fields.is_empty() {
        let keys = fields.keys().map(|k| format!("`{k}`")).collect::<Vec<_>>().join(", ");
        return Err(Error::new(Span::call_site(), format!("unexpected field(s): {keys}")));
    }

    Ok(quote! {
        CreateOfferRequest {
            credential_issuer: #credential_issuer.to_string(),
            subject_id: Some(#subject_id.to_string()),
            credential_configuration_ids: #credential_configuration_ids.iter().map(|s| s.to_string()).collect(),
            pre_authorize: #pre_authorize,
            tx_code_required: #tx_code_required,
            send_type:  #send_type,
        }
    })
}

pub fn authorization_request(input: &Data) -> Result<TokenStream> {
    // remove fields so we can error on any remaining
    let mut fields = input.fields.clone();
    let credential_issuer = fields.remove("credential_issuer").unwrap();
    let response_type = fields.remove("response_type").unwrap();
    let client_id = fields.remove("client_id").unwrap();
    let redirect_uri = fields.remove("redirect_uri").unwrap();
    let state = fields.remove("state").unwrap();
    //  let code_challenge = fields.remove("code_challenge").unwrap();
    let code_challenge_method = fields.remove("code_challenge_method").unwrap();

    // return error for any additional fields
    if !fields.is_empty() {
        let keys = fields.keys().map(|k| format!("`{k}`")).collect::<Vec<_>>().join(", ");
        return Err(Error::new(Span::call_site(), format!("unexpected field(s): {keys}")));
    }

    Ok(quote! {
        AuthorizationRequest {
            credential_issuer: #credential_issuer.to_string(),
            response_type: #response_type.to_string(),
            client_id: #client_id.to_string(),
            redirect_uri: Some(#redirect_uri.to_string()),
            state: Some(#state.to_string()),
            // code_challenge: #code_challenge.to_string(),
            code_challenge_method: #code_challenge_method.to_string(),

            // credential_configuration_ids: #credential_configuration_ids.iter().map(|s| s.to_string()).collect(),

            ..Default::default()
        }
    })
}
