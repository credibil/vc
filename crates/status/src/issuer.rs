//! # Status (Issuer)
//!
//! Traits and types for managing the publication of a credential status as an
//! issuer.

use std::future::Future;

pub use vercre_core::Quota;
pub use vercre_w3c_vc::model::CredentialStatus;

use crate::provider;

/// The `Status` trait specifies how status information can be looked up for a
/// credential.
pub trait Status: Send + Sync {
    /// Returns information on how to look up the status of a credential.
    ///
    /// A default implementation is provided that just returns `None` for cases
    /// where the issuer will not be providing a status endpoint.
    fn status(
        &self, _holder_subject: &str, _credential_identifier: &str,
    ) -> impl Future<Output = provider::Result<Option<Quota<CredentialStatus>>>> + Send {
        async { Ok(None) }
    }
}
