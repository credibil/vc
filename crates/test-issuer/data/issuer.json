{
    "credential_issuer": "http://credibil.io",
    "credential_endpoint": "http://credibil.io/credential",
    "deferred_credential_endpoint": "http://credibil.io/deferred",
    "display": {
        "name": "Credibil",
        "locale": "en-NZ"
    },
    "credential_configurations_supported": {
        "EmployeeID_JWT": {
            "format": "jwt_vc_json",
            "scope": "EmployeeIDCredential",
            "cryptographic_binding_methods_supported": [
                "did:key",
                "did:web"
            ],
            "credential_signing_alg_values_supported": [
                "ES256K",
                "EdDSA"
            ],
            "proof_types_supported": {
                "jwt": {
                    "proof_signing_alg_values_supported": [
                        "ES256K",
                        "EdDSA"
                    ]
                }
            },
            "display": [
                {
                    "name": "Employee ID",
                    "description": "Credibil employee ID credential",
                    "locale": "en-NZ",
                    "logo": {
                        "uri": "http://credibil.io/assets/employee.png",
                        "alt_text": "Employee ID Logo"
                    },
                    "text_color": "#ffffff",
                    "background_color": "#323ed2",
                    "background_image": {
                        "uri": "http://credibil.io/assets/employee-background.png",
                        "alt_text": "Employee ID Background"
                    }
                }
            ],
            "credential_definition": {
                "type": [
                    "VerifiableCredential",
                    "EmployeeIDCredential"
                ]
            },
            "claims": [
                {
                    "path": [
                        "credentialSubject",
                        "email"
                    ],
                    "display": [
                        {
                            "name": "Email",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "family_name"
                    ],
                    "mandatory": true,
                    "display": [
                        {
                            "name": "Family name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "given_name"
                    ],
                    "mandatory": true,
                    "display": [
                        {
                            "name": "Given name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address"
                    ],
                    "display": [
                        {
                            "name": "Residence",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address",
                        "street_address"
                    ],
                    "display": [
                        {
                            "name": "Street Address",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address",
                        "locality"
                    ],
                    "display": [
                        {
                            "name": "Locality",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address",
                        "region"
                    ],
                    "display": [
                        {
                            "name": "Region",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "address",
                        "country"
                    ],
                    "display": [
                        {
                            "name": "Country",
                            "locale": "en-NZ"
                        }
                    ]
                }
            ]
        },
        "Developer_JWT": {
            "format": "jwt_vc_json",
            "scope": "DeveloperCredential",
            "cryptographic_binding_methods_supported": [
                "did:key",
                "did:web"
            ],
            "credential_signing_alg_values_supported": [
                "ES256K",
                "EdDSA"
            ],
            "proof_types_supported": {
                "jwt": {
                    "proof_signing_alg_values_supported": [
                        "ES256K",
                        "EdDSA"
                    ]
                }
            },
            "display": [
                {
                    "name": "Developer",
                    "description": "Credibil certified developer credential",
                    "locale": "en-NZ",
                    "logo": {
                        "uri": "http://credibil.io/assets/developer.png",
                        "alt_text": "Developer Logo"
                    },
                    "text_color": "#ffffff",
                    "background_color": "#010100",
                    "background_image": {
                        "uri": "http://credibil.io/assets/developer-background.png",
                        "alt_text": "Developer Background"
                    }
                }
            ],
            "credential_definition": {
                "type": [
                    "VerifiableCredential",
                    "DeveloperCredential"
                ]
            },
            "claims": [
                {
                    "path": [
                        "credentialSubject",
                        "proficiency"
                    ],
                    "mandatory": true,
                    "display": [
                        {
                            "name": "Proficiency",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "family_name"
                    ],
                    "display": [
                        {
                            "name": "Family name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "credentialSubject",
                        "given_name"
                    ],
                    "display": [
                        {
                            "name": "Given name",
                            "locale": "en-NZ"
                        }
                    ]
                }
            ]
        },
        "org.iso.18013.5.1.mDL": {
            "format": "mso_mdoc",
            "doctype": "org.iso.18013.5.1.mDL",
            "scope": "DriverLicence",
            "cryptographic_binding_methods_supported": [
                "did:key",
                "did:web"
            ],
            "credential_signing_alg_values_supported": [
                "ES256K",
                "EdDSA"
            ],
            "proof_types_supported": {
                "jwt": {
                    "proof_signing_alg_values_supported": [
                        "ES256K",
                        "EdDSA"
                    ]
                }
            },
            "display": [
                {
                    "name": "Employee ID",
                    "description": "Credibil employee ID credential",
                    "locale": "en-NZ",
                    "logo": {
                        "uri": "http://credibil.io/assets/employee.png",
                        "alt_text": "Employee ID Logo"
                    },
                    "text_color": "#ffffff",
                    "background_color": "#323ed2",
                    "background_image": {
                        "uri": "http://credibil.io/assets/employee-background.png",
                        "alt_text": "Employee ID Background"
                    }
                }
            ],
            "claims": [
                {
                    "path": [
                        "org.iso.18013.5.1",
                        "given_name"
                    ],
                    "display": [
                        {
                            "name": "Given name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "org.iso.18013.5.1",
                        "family_name"
                    ],
                    "display": [
                        {
                            "name": "Family name",
                            "locale": "en-NZ"
                        }
                    ]
                },
                {
                    "path": [
                        "org.iso.18013.5.1",
                        "birth_date"
                    ],
                    "mandatory": true
                },
                {
                    "path": [
                        "org.iso.18013.5.1.aamva",
                        "organ_donor"
                    ]
                }
            ]
        }
    }
}