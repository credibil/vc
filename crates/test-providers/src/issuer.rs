use anyhow::Result;
use credibil_core::blockstore::BlockStore;
use credibil_identity::{Identity, IdentityResolver, Key, SignerExt};
use credibil_se::{Algorithm, Signer};

use crate::blockstore::Mockstore;
use crate::identity::DidIdentity;

pub const ISSUER_ID: &str = "http://credibil.io";
pub const BOB_ID: &str = "bob";
pub const CAROL_ID: &str = "carol";

pub mod data {
    pub const CLIENT: &[u8] = include_bytes!("../data/issuer/client.json");
    pub const ISSUER: &[u8] = include_bytes!("../data/issuer/issuer.json");
    pub const SERVER: &[u8] = include_bytes!("../data/issuer/server.json");
    pub const NORMAL_USER: &[u8] = include_bytes!("../data/issuer/normal-user.json");
    pub const PENDING_USER: &[u8] = include_bytes!("../data/issuer/pending-user.json");
}

#[derive(Clone)]
pub struct Issuer {
    identity: DidIdentity,
    blockstore: Mockstore,
}

impl Issuer {
    #[must_use]
    pub async fn new(owner: &str) -> Self {
        Self {
            identity: DidIdentity::new(owner).await,
            blockstore: Mockstore::new(),
        }
    }
}

impl IdentityResolver for Issuer {
    async fn resolve(&self, url: &str) -> anyhow::Result<Identity> {
        self.identity.resolve(url).await
    }
}

impl Signer for Issuer {
    async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        self.identity.try_sign(msg).await
    }

    async fn verifying_key(&self) -> Result<Vec<u8>> {
        self.identity.verifying_key().await
    }

    async fn algorithm(&self) -> Result<Algorithm> {
        Ok(self.identity.algorithm())
    }
}

impl SignerExt for Issuer {
    async fn verification_method(&self) -> Result<Key> {
        self.identity.verification_method().await
    }
}

impl BlockStore for Issuer {
    async fn put(&self, owner: &str, partition: &str, key: &str, block: &[u8]) -> Result<()> {
        self.blockstore.put(owner, partition, key, block).await
    }

    async fn get(&self, owner: &str, partition: &str, key: &str) -> Result<Option<Vec<u8>>> {
        self.blockstore.get(owner, partition, key).await
    }

    async fn delete(&self, owner: &str, partition: &str, key: &str) -> Result<()> {
        self.blockstore.delete(owner, partition, key).await
    }

    async fn purge(&self, _owner: &str, _partition: &str) -> Result<()> {
        unimplemented!()
    }
}
