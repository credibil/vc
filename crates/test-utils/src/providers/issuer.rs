use std::collections::HashMap;

use anyhow::anyhow;
use openid4vc::issuance::Issuer;
use provider::Result;

#[derive(Default, Clone, Debug)]
pub struct Store {
    issuers: HashMap<String, Issuer>,
}

impl Store {
    pub fn new() -> Self {
        let issuer = Issuer::sample();

        Self {
            issuers: HashMap::from([
                ("http://localhost:8080".into(), issuer.clone()),
                (issuer.credential_issuer.clone(), issuer),
            ]),
        }
    }

    pub fn get(&self, issuer_id: &str) -> Result<Issuer> {
        let Some(issuer) = self.issuers.get(issuer_id) else {
            return Err(anyhow!("issuer not found"));
        };
        Ok(issuer.clone())
    }
}
