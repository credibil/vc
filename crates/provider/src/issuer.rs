#![allow(unused)]

use anyhow::Result;
use credibil_did::{DidResolver, Document};
use credibil_infosec::{Algorithm, Signer};
use credibil_vc::BlockStore;
use credibil_vc::status::issuer::Status;

use crate::blockstore::Mockstore;
use crate::keystore::Keyring;

pub const ISSUER_ID: &str = "http://credibil.io";
pub const BOB_ID: &str = "bob";
pub const CAROL_ID: &str = "carol";

pub mod data {
    pub const CLIENT: &[u8] = include_bytes!("../data/issuer/client.json");
    pub const ISSUER: &[u8] = include_bytes!("../data/issuer/issuer.json");
    pub const SERVER: &[u8] = include_bytes!("../data/issuer/server.json");
    pub const NORMAL_USER: &[u8] = include_bytes!("../data/issuer/normal-user.json");
    pub const PENDING_USER: &[u8] = include_bytes!("../data/issuer/pending-user.json");
}

#[derive(Clone, Debug)]
pub struct Issuer {
    keyring: Keyring,
    blockstore: Mockstore,
}

impl Issuer {
    #[must_use]
    pub fn new() -> Self {
        Self {
            keyring: Keyring::new(),
            blockstore: Mockstore::new(),
        }
    }
}

impl DidResolver for Issuer {
    async fn resolve(&self, url: &str) -> anyhow::Result<Document> {
        self.keyring.resolve(url).await
    }
}

impl Signer for Issuer {
    async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        self.keyring.try_sign(msg).await
    }

    async fn verifying_key(&self) -> Result<Vec<u8>> {
        self.keyring.verifying_key().await
    }

    fn algorithm(&self) -> Algorithm {
        self.keyring.algorithm()
    }

    async fn verification_method(&self) -> Result<String> {
        self.keyring.verification_method().await
    }
}

impl BlockStore for Issuer {
    async fn put(&self, owner: &str, partition: &str, key: &str, block: &[u8]) -> Result<()> {
        self.blockstore.put(owner, partition, key, block).await
    }

    async fn get(&self, owner: &str, partition: &str, key: &str) -> Result<Option<Vec<u8>>> {
        self.blockstore.get(owner, partition, key).await
    }

    async fn delete(&self, owner: &str, partition: &str, key: &str) -> Result<()> {
        self.blockstore.delete(owner, partition, key).await
    }

    async fn purge(&self, _owner: &str, _partition: &str) -> Result<()> {
        unimplemented!()
    }
}

impl Status for Issuer {}
