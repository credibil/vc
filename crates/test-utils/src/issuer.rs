use chrono::{DateTime, Utc};
use serde::de::DeserializeOwned;
use serde::Serialize;
use vercre_datasec::{Algorithm, Decryptor, Encryptor, SecOps, Signer};
use vercre_did::{DidResolver, Document};
use vercre_openid::issuer::{
    Dataset, Client, Issuer, Metadata, Result, Server, StateStore, Subject,
};

use crate::store::keystore::IssuerKeystore;
use crate::store::{issuance, resolver, state};

pub const CREDENTIAL_ISSUER: &str = "http://vercre.io";
pub const CLIENT_ID: &str = "96bfb9cb-0513-7d64-5532-bed74c48f9ab";
pub const NORMAL_USER: &str = "normal_user";
pub const PENDING_USER: &str = "pending_user";

#[derive(Default, Clone, Debug)]
pub struct Provider {
    pub client: issuance::ClientStore,
    pub issuer: issuance::IssuerStore,
    pub server: issuance::ServerStore,
    pub subject: issuance::DatasetStore,
    pub state: state::Store,
}

impl Provider {
    #[must_use]
    pub fn new() -> Self {
        Self {
            client: issuance::ClientStore::new(),
            issuer: issuance::IssuerStore::new(),
            server: issuance::ServerStore::new(),
            subject: issuance::DatasetStore::new(),
            state: state::Store::new(),
        }
    }
}

impl vercre_openid::issuer::Provider for Provider {}

impl Metadata for Provider {
    async fn client(&self, client_id: &str) -> Result<Client> {
        self.client.get(client_id)
    }

    async fn register(&self, client: &Client) -> Result<Client> {
        self.client.add(client)
    }

    async fn issuer(&self, issuer_id: &str) -> Result<Issuer> {
        self.issuer.get(issuer_id)
    }

    async fn server(&self, server_id: &str) -> Result<Server> {
        self.server.get(server_id)
    }
}

impl Subject for Provider {
    /// Authorize issuance of the specified credential for the holder.
    async fn authorize(
        &self, subject_id: &str, credential_configuration_id: &str,
    ) -> Result<Vec<String>> {
        self.subject.authorize(subject_id, credential_configuration_id)
    }

    async fn dataset(&self, subject_id: &str, credential_identifier: &str) -> Result<Dataset> {
        self.subject.dataset(subject_id, credential_identifier)
    }
}

impl StateStore for Provider {
    async fn put(&self, key: &str, state: impl Serialize, dt: DateTime<Utc>) -> Result<()> {
        self.state.put(key, state, dt)
    }

    async fn get<T: DeserializeOwned>(&self, key: &str) -> Result<T> {
        self.state.get(key)
    }

    async fn purge(&self, key: &str) -> Result<()> {
        self.state.purge(key)
    }
}

impl DidResolver for Provider {
    async fn resolve(&self, url: &str) -> anyhow::Result<Document> {
        resolver::resolve_did(url).await
    }
}

struct IssuerSec(IssuerKeystore);

impl SecOps for Provider {
    fn signer(&self, _identifier: &str) -> anyhow::Result<impl Signer> {
        Ok(IssuerSec(IssuerKeystore {}))
    }

    fn encryptor(&self, _identifier: &str) -> anyhow::Result<impl Encryptor> {
        Ok(IssuerSec(IssuerKeystore {}))
    }

    fn decryptor(&self, _identifier: &str) -> anyhow::Result<impl Decryptor> {
        Ok(IssuerSec(IssuerKeystore {}))
    }
}

impl Signer for IssuerSec {
    fn algorithm(&self) -> Algorithm {
        self.0.algorithm()
    }

    fn verification_method(&self) -> String {
        self.0.verification_method()
    }

    async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        self.0.try_sign(msg)
    }
}

impl Encryptor for IssuerSec {
    async fn encrypt(&self, _plaintext: &[u8], _recipient_public_key: &[u8]) -> Result<Vec<u8>> {
        todo!()
    }

    fn public_key(&self) -> Vec<u8> {
        todo!()
    }
}

impl Decryptor for IssuerSec {
    async fn decrypt(&self, _ciphertext: &[u8], _sender_public_key: &[u8]) -> Result<Vec<u8>> {
        todo!()
    }
}
