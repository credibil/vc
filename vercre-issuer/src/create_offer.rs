//! # Invoke Endpoint
//!
//! The Invoke endpoint is used to initiate Pre-authorized credential issuance flow.
//! Credential Issuers can use this endpoint to generate a Credential Offer which can
//! be used to initiate issuance with a Wallet.
//!
//! When a Credential Issuer is already interacting with a user and wishes to initate a
//! Credential issuance, they can 'send' the user's Wallet a Credential Offer.
//!
//! The diagram illustrates this Credential Issuer initiated flow:
//!
//! ```text
//! +--------------+   +-----------+                                    +-------------------+
//! | User         |   |   Wallet  |                                    | Credential Issuer |
//! +--------------+   +-----------+                                    +-------------------+
//!         |                |                                                    |
//!         |                |  (1) User provides  information required           |
//!         |                |      for the issuance of a certain Credential      |
//!         |-------------------------------------------------------------------->|
//!         |                |                                                    |
//!         |                |  (2) Credential Offer (Pre-Authorized Code)        |
//!         |                |<---------------------------------------------------|
//!         |                |  (3) Obtains Issuer's Credential Issuer metadata   |
//!         |                |<-------------------------------------------------->|
//!         |   interacts    |                                                    |
//!         |--------------->|                                                    |
//!         |                |                                                    |
//!         |                |  (4) Token Request (Pre-Authorized Code, pin)      |
//!         |                |--------------------------------------------------->|
//!         |                |      Token Response (access_token)                 |
//!         |                |<---------------------------------------------------|
//!         |                |                                                    |
//!         |                |  (5) Credential Request (access_token, proof(s))   |
//!         |                |--------------------------------------------------->|
//!         |                |      Credential Response                           |
//!         |                |      (credential(s))                               |
//!         |                |<---------------------------------------------------|
//! ```
//!
//! While JSON-based, the Offer can be sent to the Wallet's Credential Offer Handler URL
//! as an HTTP GET request, an HTTP redirect, or a QR code.
//!
//! Below is a non-normative example of a Credential Offer Object for a Pre-Authorized
//! Code Stage (with a credential type reference):
//!
//! ```json
//! {
//!     "credential_issuer": "https://credential-issuer.example.com",
//!     "credential_configuration_ids": [
//!         "UniversityDegree_LDP_VC"
//!     ],
//!     "grants": {
//!         "urn:ietf:params:oauth:grant-type:pre-authorized_code": {
//!             "pre-authorized_code": "adhjhdjajkdkhjhdj",
//!             "tx_code": {
//!                 "input_mode":"numeric",
//!                 "length":6,
//!                 "description":"Please provide the one-time code that was sent via e-mail"
//!             }
//!        }
//!     }
//! }
//! ```
//!
//! See <https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-credential-offer-endpoint>

use std::collections::HashMap;

use chrono::Utc;
use tracing::instrument;
use vercre_core::gen;
use vercre_openid::issuer::{
    AuthorizationCodeGrant, CreateOfferRequest, CreateOfferResponse, CredentialOffer, Grants,
    Metadata, OfferType, PreAuthorizedCodeGrant, Provider, SendType, StateStore, Subject, TxCode,
};
use vercre_openid::{Error, Result};

use crate::state::{Expire, Offer, PreAuthorization, Stage, State};

/// Invoke request handler generates and returns a Credential Offer.
///
/// # Errors
///
/// Returns an `OpenID4VP` error if the request is invalid or if the provider is
/// not available.
#[instrument(level = "debug", skip(provider))]
pub async fn create_offer(
    provider: impl Provider, request: CreateOfferRequest,
) -> Result<CreateOfferResponse> {
    verify(&provider, &request).await?;
    process(&provider, request).await
}

async fn verify(provider: &impl Provider, request: &CreateOfferRequest) -> Result<()> {
    tracing::debug!("create_offer::verify");

    let issuer_meta = Metadata::issuer(provider, &request.credential_issuer)
        .await
        .map_err(|e| Error::ServerError(format!("metadata issue: {e}")))?;

    // credential_issuer required
    if request.credential_issuer.is_empty() {
        return Err(Error::InvalidRequest("no credential_issuer specified".into()));
    };

    // credentials required
    if request.credential_configuration_ids.is_empty() {
        return Err(Error::InvalidRequest("no credentials requested".into()));
    };

    // requested credential is supported
    for cred_id in &request.credential_configuration_ids {
        if !issuer_meta.credential_configurations_supported.contains_key(cred_id) {
            return Err(Error::UnsupportedCredentialType(
                "requested credential is unsupported".into(),
            ));
        };
    }

    // subject_id is required
    if request.subject_id.is_none() {
        return Err(Error::InvalidRequest("no subject_id specified".into()));
    };

    Ok(())
}

// Process the request.
async fn process(
    provider: &impl Provider, request: CreateOfferRequest,
) -> Result<CreateOfferResponse> {
    tracing::debug!("create_offer::process");

    let credential_offer = credential_offer(&request);
    let tx_code =
        if request.pre_authorize && request.tx_code_required { Some(gen::tx_code()) } else { None };
    let credentials = authorized_credentials(provider, &request).await?;

    // ------------------------------------------------------------------------
    // save state
    // ------------------------------------------------------------------------
    let state_key = state_key(&credential_offer)?;

    let state_stage = if request.pre_authorize && request.send_type == SendType::ByVal {
        Stage::PreAuthorized(PreAuthorization {
            credentials: credentials.clone(),
            tx_code: tx_code.clone(),
        })
    } else {
        Stage::Offered(Offer {
            credential_offer: credential_offer.clone(),
            credentials: credentials.clone(),
            tx_code: tx_code.clone(),
        })
    };
    let state = State {
        expires_at: Utc::now() + Expire::Authorized.duration(),
        subject_id: request.subject_id.clone(),
        stage: state_stage,
    };
    StateStore::put(provider, &state_key, &state, state.expires_at)
        .await
        .map_err(|e| Error::ServerError(format!("issue saving state: {e}")))?;
    // ------------------------------------------------------------------------

    // TODO: use unique id rather than pre_authorized_code in URI
    let offer_type = if request.send_type == SendType::ByRef {
        // TODO: use unique id rather than pre_authorized_code in URI
        OfferType::Uri(format!("{}/credential_offer/{state_key}", request.credential_issuer))
    } else {
        OfferType::Object(credential_offer)
    };

    Ok(CreateOfferResponse { offer_type, tx_code })
}

fn state_key(credential_offer: &CredentialOffer) -> Result<String> {
    // get pre-authorized code as state key
    let Some(grants) = &credential_offer.grants else {
        return Err(Error::ServerError("no grants".into()));
    };

    if let Some(pre_auth_code) = &grants.pre_authorized_code {
        return Ok(pre_auth_code.pre_authorized_code.clone());
    };

    if let Some(authorization_code) = &grants.authorization_code {
        let Some(issuer_state) = &authorization_code.issuer_state else {
            return Err(Error::ServerError("no issuer_state".into()));
        };
        return Ok(issuer_state.clone());
    };

    Err(Error::ServerError("no grants".into()))
}

// Create CredentialOffer
fn credential_offer(request: &CreateOfferRequest) -> CredentialOffer {
    let grants = if request.pre_authorize {
        let tx_code_def = if request.tx_code_required {
            Some(TxCode {
                input_mode: Some("numeric".into()),
                length: Some(6),
                description: Some("Please provide the one-time code received".into()),
            })
        } else {
            None
        };

        Grants {
            authorization_code: None,
            pre_authorized_code: Some(PreAuthorizedCodeGrant {
                pre_authorized_code: gen::auth_code(),
                tx_code: tx_code_def,
                authorization_server: None,
            }),
        }
    } else {
        Grants {
            authorization_code: Some(AuthorizationCodeGrant {
                issuer_state: Some(gen::state_key()),
                authorization_server: None,
            }),
            pre_authorized_code: None,
        }
    };

    CredentialOffer {
        credential_issuer: request.credential_issuer.clone(),
        credential_configuration_ids: request.credential_configuration_ids.clone(),
        grants: Some(grants),
    }
}

async fn authorized_credentials(
    provider: &impl Provider, request: &CreateOfferRequest,
) -> Result<HashMap<String, String>> {
    let Some(subject_id) = &request.subject_id else {
        return Err(Error::InvalidRequest(
            "`subject_id` must be set for pre-authorized offers".into(),
        ));
    };

    let mut credentials = HashMap::new();

    for config_id in request.credential_configuration_ids.clone() {
        let identifiers = Subject::authorize(provider, subject_id, &config_id, None)
            .await
            .map_err(|e| Error::ServerError(format!("issue authorizing holder: {e}")))?;

        for identifier in &identifiers {
            credentials.insert(
                identifier.clone(),
                config_id.clone(),
                // AuthorizedCredential {
                //     credential_identifier: identifier.clone(),
                //     credential_configuration_id: config_id.clone(),
                // },
            );
        }
    }

    Ok(credentials)
}

#[cfg(test)]
mod tests {
    use assert_let_bind::assert_let;
    use insta::assert_yaml_snapshot as assert_snapshot;
    use vercre_macros::create_offer_request;
    use vercre_test_utils::issuer::{Provider, CREDENTIAL_ISSUER, NORMAL_USER};
    use vercre_test_utils::snapshot;

    use super::*;

    #[tokio::test]
    async fn pre_authorized() {
        vercre_test_utils::init_tracer();
        snapshot!("");

        let provider = Provider::new();

        // create offer to 'send' to the app
        let request = create_offer_request!({
            "credential_issuer": CREDENTIAL_ISSUER,
            "credential_configuration_ids": ["EmployeeID_JWT"],
            "subject_id": NORMAL_USER,
            "pre_authorize": true,
            "tx_code_required": true,
            "send_type": SendType::ByVal,
        });

        let response = create_offer(provider.clone(), request).await.expect("response is ok");

        assert_snapshot!("create_offer:pre-authorized:response", &response, {
            ".credential_offer.grants.authorization_code.issuer_state" => "[state]",
            ".credential_offer.grants[\"urn:ietf:params:oauth:grant-type:pre-authorized_code\"][\"pre-authorized_code\"]" => "[pre-authorized_code]",
            ".tx_code" => "[tx_code]"
        });

        // check redacted fields
        let OfferType::Object(offer) = response.offer_type else {
            panic!("expected CredentialOfferType::Object");
        };
        assert_let!(Some(grants), &offer.grants);
        assert_let!(Some(pre_auth_code), &grants.pre_authorized_code);
        assert!(grants.pre_authorized_code.is_some());

        // compare response with saved state
        let pre_auth_code = &pre_auth_code.pre_authorized_code; //as_ref().expect("has state");
        let state = StateStore::get::<State>(&provider, pre_auth_code).await.expect("state exists");

        assert_snapshot!("create_offer:pre-authorized:state", &state, {
            ".expires_at" => "[expires_at]",
            ".stage.tx_code" => "[tx_code]"
        });

        assert_let!(Stage::PreAuthorized(auth_state), &state.stage);
        assert_eq!(auth_state.tx_code, response.tx_code);
    }
}
