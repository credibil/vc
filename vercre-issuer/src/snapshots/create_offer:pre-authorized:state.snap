---
source: vercre-issuer/src/create_offer.rs
assertion_line: 313
expression: "&state"
---
expires_at: "[expires_at]"
subject_id: normal_user
stage:
  pre_authorized:
    credentials:
      PHLEmployeeID: EmployeeID_JWT
    tx_code: "[tx_code]"
